File: C++/lab1.cpp
#include <iostream>
#include <vector>

class Quest1
{
    public:
        int _number;
        int _index;
        int _count = 0;
        int _digit;
        std::vector<int> _divider_arr;
        static void setupForConsole()
        {
            Quest1 q1;
            q1.inputNumberAndIndex();
            q1.digitCount();
            q1.sliceDigitByIndex();
            std::cout << "Число по индексу " << q1._index << ": " << q1._digit << std::endl;

            q1.inputDividers();
            q1.checkDivisibility();
        }

        void inputNumberAndIndex()
        {
            std::cout << "Введите число: ";
            std::cin >> _number;
            std::cout << "Введите индекс: ";
            std::cin >> _index;
        }

        void inputDividers()
        {
            int count;
            std::cout << "Введите количество делителей: ";
            std::cin >> count;
            int tmp_div;

            for (int i=0; i < count; i++)
            {
                std::cout << "Введите " << i+1 << " делитель: ";
                std::cin >> tmp_div;
                _divider_arr.push_back(tmp_div);
            }
        }
        void digitCount()
        {
            int tmp_number;
            tmp_number = _number;
            while (tmp_number > 0)
            {
                tmp_number /= 10;
                _count++;
            }
        }

        void sliceDigitByIndex()
        {
            int tmp_number;
            tmp_number = _number;
            if (_count >= _index)
            {
                for (int i=0; i < _count - _index; i++)
                {
                    tmp_number /= 10;
                }
            }
            _digit = tmp_number % 10;
        }

        void checkDivisibility()
        {
            for (int divider : _divider_arr)
            {
                if (_number % divider == 0)
                {
                    std::cout << _number << " делится на " << divider << std::endl;
                }
                else
                {
                    std::cout << _number << " не делится на " << divider << std::endl;
                }
            }
        }
};
class Quest2
{
public:
    int _start=10, _stop=100;
    int _array1[90];
    int _array2[90];
    int _size=90;
    std::vector<int> _vector_2_8;
    std::vector<int> _vector_3_4;
    static void setupForConsole()
    {
        Quest2 q2;
        q2.fillArrays();
        q2.findLastDigit();
        std::cout << "При умножении на 2 - конец 8:" << std::endl;
        for (int number: q2._vector_2_8)
        {
            std::cout << number << " ";
        }
        std::cout << "\n" <<std::endl;
        std::cout << "При умножении на 3 - конец 4: " << std::endl;
        for (int number: q2._vector_3_4)
        {
            std::cout << number << " ";
        }
    }
    void fillArrays()
    {
        for (int i=0; i < _size; i++)
        {
            _array1[i] = (i + _start);
            _array2[i] = (i + _start);
        }
    }
    void findLastDigit()
    {
        for (int i=0; i < _size; i++)
        {
            if (_array1[i]*2 % 10 == 8)
            {
                _vector_2_8.push_back(_array1[i]);
            }
            if (_array2[i]*3 % 10 == 4)
            {
                _vector_3_4.push_back(_array2[i]);
            }
        }
    }
};

int main()
{
    Quest1::setupForConsole();
    Quest2::setupForConsole();
    return 0;
}

File: Python/lab1.py
class Quest1:
    def __init__(self):
        self.number = 0
        self.index = 0
        self.count = 0
        self.digit = 0
        self.divider_list = list()
        self.divider_dict = dict()

    @staticmethod
    def setup_for_console():
        q1 = Quest1()
        q1.input_number_and_index()
        q1.digit_count()
        q1.slice_digit_by_index()
        print(f"Цифра: {q1.digit}")
        q1.input_dividers()
        q1.check_divisibility()
        print(f"Делимости:\n{q1.divider_dict}")

    def input_number_and_index(self):
        self.number = int(input("Введите число: "))
        self.index = int(input("Введите индекс цифры: "))

    def input_dividers(self):
        count = int(input("Введите количество делителей: "))
        for i in range(count):
            self.divider_list.append(int(input(f"Введите {i+1} делитель: ")))

    def digit_count(self):
        tmp_number = self.number
        while tmp_number > 0:
            tmp_number //= 10
            self.count += 1

    def slice_digit_by_index(self):
        tmp_number = self.number
        if self.count > self.index:
            for i in range(self.count - self.index):
                tmp_number //= 10

        self.digit = tmp_number % 10

    def check_divisibility(self):
        for divider in self.divider_list:
            self.divider_dict[f"{divider}"] = not bool(self.number % divider)


class Quest2:
    def __init__(self, start=10, stop=100):
        self.list1 = [i for i in range(start, stop)]
        self.list2 = [i for i in range(start, stop)]
        self.list_2_8 = list()
        self.list_3_4 = list()

    @staticmethod
    def setup_for_console():
        q2 = Quest2()
        q2.find_last_digit()
        print(f"При умножении на 2 - конец 8: {q2.list_2_8}\n"
              f"При умножении на 3 - конец 4: {q2.list_3_4}\n"
              f"{len(q2.list1)}")

    def find_last_digit(self):
        for i in range(len(self.list1)):
            if self.list1[i]*2 % 10 == 8:
                self.list_2_8.append(self.list1[i])
            if self.list2[i]*3 % 10 == 4:
                self.list_3_4.append((self.list2[i]))


if __name__ == "__main__":
    Quest1.setup_for_console()
    Quest2.setup_for_console()
